src/uct/ib/base/ib_device.c

```diff
@@ -482,24 +482,77 @@ void uct_ib_handle_async_event(uct_ib_device_t *dev, uct_ib_async_event_t *event
     ucs_log(level, "IB Async event on %s: %s", uct_ib_device_name(dev), event_info);
 }
 
+static ucs_status_t uct_ib_device_get_path_buffer(uct_ib_device_t *dev,
+                                                  char *path_buffer)
+{
+    char *resolved_path;
+
+    resolved_path = realpath(dev->ibv_context->device->ibdev_path, path_buffer);
+    if (resolved_path == NULL) {
+        return UCS_ERR_IO_ERROR;
+    }
+
+    /* Make sure there is "/infiniband/" substring in path_buffer */
+    if (strstr(path_buffer, "/infiniband/") == NULL) {
+        return UCS_ERR_IO_ERROR;
+    }
+
+    return UCS_OK;
+}
+
+static ucs_status_t uct_ib_device_get_ids_from_path(const char *path,
+                                                    long *vendor_id,
+                                                    long *device_id)
+{
+    ucs_status_t status;
+
+    status = ucs_read_file_number(vendor_id, 1, "%s/%s", path, "vendor");
+    if (status != UCS_OK) {
+        return status;
+    }
+
+    status = ucs_read_file_number(device_id, 1, "%s/%s", path, "device");
+    if (status != UCS_OK) {
+        return status;
+    }
+
+    return UCS_OK;
+}
+
 static void uct_ib_device_get_ids(uct_ib_device_t *dev)
 {
     long vendor_id, device_id;
+    char *ids_path;
+    char path_buffer[PATH_MAX];
+    ucs_status_t status;
+    /* PF: 0, SF: 3 */
+    int nums_layers[2] = {2, 3};
+    int i;
+
+    /* SF: realpath name is of form /sys/devices/.../0000:03:00.0/<UUID>/infiniband/mlx5_0 */
+    /* PF: realpath name is of form /sys/devices/.../0000:03:00.0/infiniband/mlx5_0 */
+
+    status = uct_ib_device_get_path_buffer(dev, path_buffer);
+    if (status == UCS_OK) {
+        for (i = 0; i < 2; i++) {
+            ids_path = ucs_dirname(path_buffer, nums_layers[i]);
+            if (ids_path != NULL) {
+                if (uct_ib_device_get_ids_from_path(ids_path, &vendor_id, &device_id) == UCS_OK) {
```

Comment:
maybe pass &dev->pci_id.vendor and &dev->pci_id.device?

Reply:
@yosefe Can't pass them directly since the variable types are different with the function parameters. Have changed to another way.